### ISSUE TITLE
Handling calls to NODE that uses stateful layers

### ISSUE THREAD

**Issue by adam-hartshorne** [pnkraemer/probdiffeq#801]
I know that probdiffeq handles the passing of arguments to the vector field function e.g.

```
def vf(y, *, t, p):
    coeff = p['coeff']
    return coeff * y
```
called with something like

`sol = ivpsolve.solve_fixed_grid(lambda *a, **kw: vf(*a, **kw, p=f_args), init, grid=grid, solver=solver_ts0)`

Now in equinox if we want to use BatchNormLayer in our NN, you have to pass around this state argument that gets updated everytime the NN is called e.g.

https://docs.kidger.site/equinox/examples/stateful/

is it possible to pass in the state to the vector field argument in a way that allows it to be updated correctly? I presume the p arg to pass in additional ODE parameters is assuming that such parameters are not changing. 

Basically I need to be able to do something like this
```
def vf(y, *, t, state):
    yt, state = self.ode_func(y, state)
    return yt, state
```

**Comment by pnkraemer** [comment]
That's a good question... I don't know, let's try to find out.

Do I understand correctly that you want the parameter to change with every function call? Have you considered stacking "y" and "state" into a single vector? (presumably, after calling something like jax.flatten_util.ravel_pytree). In this case, the ODE could be solved like a normal ODE.

**Comment by adam-hartshorne** [comment]
> Do I understand correctly that you want the parameter to change with every function call? 

Yes, this state stores a special copy of the NN structure with running averages for mean / std of the weights e.g. if I print out the state structure for my simple test you get something like this,

State(
  0x7a81bab000d0=bool[],
  0x7a81bab000f0=(f32[256], f32[256]),
  0x7a81bab00110=bool[],
  0x7a81bab00130=(f32[256], f32[256]),
  0x7a81bab00150=bool[],
  0x7a81bab00170=(f32[256], f32[256]),
  0x7a81bab00190=bool[],
  0x7a81bab001b0=(f32[256], f32[256])
)

so you can't just flatten it. Also, we don't want the ODE solver to try and "solve" for it in anyway, just to pass around this modified state.

**Comment by pnkraemer** [comment]
I see, thanks for iterating. Do you know whether (and if so, how) non-probabilistic solvers handle this case?

**Comment by adam-hartshorne** [comment]
I don't believe Diffrax can. It has a similar method for passing static args. I know its a very strange request. 

Obviously NODEs, normally are very straightforward MLP with something like GELU or Tanh activation functions. The big problem with such approaches are heavily biased to learning only very smooth flows as these activation functions are heavily bias against learning high-frequency "features". In many cases this is what you want. In a recent paper, I successfully used a version of SIREN activations to allow some more fine-grain deviations in the flow, but you have to be rather careful over initialisation and training them.  

I have just read this paper ( https://arxiv.org/pdf/2407.17834 ), which claims you can overcome this spectrum bias in coordinate NNs by simply using normalisations such as BatchNorm layers. So I was interested to see if you applied it to a NODE what would happen.

**Comment by pnkraemer** [comment]
I think I understand what you mean. I am wondering whether what you're describing is even a well-defined ODE (as in, one that admits a unique, differentiable solution) because the vector field changes so frequently. I would speculate that it isn't, but I might be wrong. This concern is important because if what we're discussing is not a well-posed ODE problem, then we're not really discussing ODE solvers anymore... 

What do you think? 

By the way, if you think that this issue board is not the right place to discuss, feel free to shoot me an email :) (I think we might be entering research-territory here).

**Comment by adam-hartshorne** [comment]
I will shoot you an email.

**Comment by pnkraemer** [comment]
Cool! Let's close this issue and reopen if necessary.
